File: srv_erp/srv_erp/report/items_delivered_in_date_range/items_delivered_in_date_range.js
// Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Items Delivered in Date Range"] = {
	onload: async function(report) {
		await set_finished_warehouse(report);
	},
	filters: [
		{
			fieldname: "pending_only",
			label: __("Pending Only"),
			fieldtype: "Check",
			default: 1,
			description: __("Show only delivered items whose Sales Order quantity is still pending."),
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "range",
			label: __("Range"),
			fieldtype: "Select",
			options: ["Daily", "Weekly", "Monthly", "Quarterly", "Yearly"],
			default: "Daily",
			reqd: 1,
			on_change: function() {
				let range = frappe.query_report.get_filter_value('range');
				let today = frappe.datetime.get_today();
				let from_date = today;
				
				if (range === 'Weekly') {
					from_date = frappe.datetime.add_days(today, -7);
				} else if (range === 'Monthly') {
					from_date = frappe.datetime.add_months(today, -1);
				} else if (range === 'Quarterly') {
					from_date = frappe.datetime.add_months(today, -3);
				} else if (range === 'Yearly') {
					from_date = frappe.datetime.add_months(today, -12);
				}
				
				frappe.query_report.set_filter_value('from_date', from_date);
				frappe.query_report.set_filter_value('to_date', today);
			}
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "customer_group",
			label: __("Customer Group"),
			fieldtype: "Link",
			options: "Customer Group",
		},
		{
			fieldname: "territory",
			label: __("Territory"),
			fieldtype: "Link",
			options: "Territory",
		},
		{
			fieldname: "sales_person",
			label: __("Sales Person"),
			fieldtype: "Link",
			options: "Sales Person",
		},
		{
			fieldname: "item_code",
			label: __("Item"),
			fieldtype: "Link",
			options: "Item",
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
		},
		{
			fieldname: "brand",
			label: __("Brand"),
			fieldtype: "Link",
			options: "Brand",
		},
		{
			fieldname: "include_uom",
			label: __("Include UOM"),
			fieldtype: "Link",
			options: "UOM",
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
		},
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
			on_change: async function() {
				await set_finished_warehouse(frappe.query_report, true);
			},
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
			reqd: 1,
			get_query() {
				return {
					filters: {
						company: frappe.query_report.get_filter_value("company"),
						is_group: 0,
						disabled: 0,
					},
				};
			},
		},
	],
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (!data) {
			return value;
		}

		const status_fields = [
			"stock_ordered_qty",
			"stock_delivered_qty",
			"stock_available_qty",
			"stock_shortfall_qty",
		];
		const shortage = flt(data.stock_shortfall_qty);
		const ordered = flt(data.stock_ordered_qty);
		const delivered = flt(data.stock_delivered_qty);
		const stock = flt(data.stock_available_qty);

		if (column.fieldname === "item_name") {
			const color = shortage > 0 ? "#b3261e" : "#137333";
			return `<span style="border-left:3px solid ${color};padding-left:6px">${value}</span>`;
		}
		if (!status_fields.includes(column.fieldname)) {
			return value;
		}

		let foreground = "#1a73e8";
		let background = "#e8f0fe";
		if (column.fieldname === "stock_shortfall_qty") {
			foreground = shortage > 0 ? "#b3261e" : "#137333";
			background = shortage > 0 ? "#fce8e6" : "#e6f4ea";
		} else if (column.fieldname === "stock_available_qty") {
			const stock_is_sufficient = stock >= Math.max(ordered - delivered, 0);
			foreground = stock_is_sufficient ? "#137333" : "#b3261e";
			background = stock_is_sufficient ? "#e6f4ea" : "#fce8e6";
		} else if (column.fieldname === "stock_delivered_qty") {
			foreground = delivered > 0 ? "#137333" : "#b3261e";
			background = delivered > 0 ? "#e6f4ea" : "#fce8e6";
		}

		return `<span style="display:block;margin:-2px -4px;padding:2px 4px;border-radius:3px;color:${foreground};background:${background};font-weight:700">${value}</span>`;
	},
};

async function set_finished_warehouse(report, replace_existing = false) {
	const warehouse_filter = report.get_filter("warehouse");
	if (!warehouse_filter || (!replace_existing && warehouse_filter.get_value())) {
		return;
	}

	const company = report.get_filter_value("company");
	if (!company) {
		return;
	}

	const warehouses = await frappe.db.get_list("Warehouse", {
		filters: {
			company,
			warehouse_name: "Finished Goods",
			is_group: 0,
			disabled: 0,
		},
		fields: ["name", "warehouse_name"],
		limit: 1,
	});
	const warehouse = warehouses[0];
	await warehouse_filter.set_value(warehouse ? warehouse.name : "");
}

File: srv_erp/srv_erp/report/items_delivered_in_date_range/items_delivered_in_date_range.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from dateutil.relativedelta import relativedelta
from frappe import _
from frappe.utils import cint, getdate

from srv_erp.srv_erp.report.hierarchical_filters import get_descendant_condition
from srv_erp.srv_erp.report.uom_utils import add_selected_uom_columns

STOCK_ORDERED_QTY_SQL = "COALESCE(soi.stock_qty, dni.stock_qty)"
STOCK_DELIVERED_QTY_SQL = (
	"COALESCE("
	"soi.delivered_qty * COALESCE(NULLIF(soi.conversion_factor, 0), 1), "
	"dni.stock_qty"
	")"
)


def execute(filters=None):
	filters = frappe._dict(filters or {})
	set_default_warehouse(filters)
	validate_warehouse(filters)
	columns = get_columns()
	data = get_data(filters)
	add_selected_uom_columns(columns, data, filters.get("include_uom"))
	return columns, data


def get_columns():
	return [
		{
			"label": _("Item Code"),
			"fieldname": "item_code",
			"fieldtype": "Link",
			"options": "Item",
			"width": 120,
		},
		{
			"label": _("Item Name"),
			"fieldname": "item_name",
			"fieldtype": "Data",
			"width": 160,
		},
		{"label": _("Ordered"), "fieldname": "stock_ordered_qty", "fieldtype": "Float", "width": 100},
		{"label": _("Delivered"), "fieldname": "stock_delivered_qty", "fieldtype": "Float", "width": 100},
		{"label": _("Stock"), "fieldname": "stock_available_qty", "fieldtype": "Float", "width": 100},
		{
			"label": _("To Produce"),
			"fieldname": "stock_shortfall_qty",
			"fieldtype": "Float",
			"width": 110,
			"description": _("Ordered - Delivered - Stock (minimum 0)"),
		},
		{"label": _("Stock UOM"), "fieldname": "production_uom", "fieldtype": "Link", "options": "UOM", "width": 90},
		{
			"label": _("Item Group"),
			"fieldname": "item_group",
			"fieldtype": "Link",
			"options": "Item Group",
			"width": 120,
		},
		{
			"label": _("Brand"),
			"fieldname": "brand",
			"fieldtype": "Link",
			"options": "Brand",
			"width": 120,
		},
		{
			"label": _("Customer"),
			"fieldname": "customer",
			"fieldtype": "Link",
			"options": "Customer",
			"width": 140,
		},
		{
			"label": _("Customer Name"),
			"fieldname": "customer_name",
			"fieldtype": "Data",
			"width": 160,
		},
		{
			"label": _("Customer Group"),
			"fieldname": "customer_group",
			"fieldtype": "Link",
			"options": "Customer Group",
			"width": 130,
		},
		{
			"label": _("Territory"),
			"fieldname": "territory",
			"fieldtype": "Link",
			"options": "Territory",
			"width": 120,
		},
		{
			"label": _("Sales Person"),
			"fieldname": "sales_person",
			"fieldtype": "Data",
			"width": 140,
		},
		{
			"label": _("Posting Date"),
			"fieldname": "posting_date",
			"fieldtype": "Date",
			"width": 100,
		},
		{
			"label": _("Qty Delivered"),
			"fieldname": "qty",
			"fieldtype": "Float",
			"width": 100,
		},
		{
			"label": _("Qty Delivered UOM"),
			"fieldname": "uom_qty",
			"fieldtype": "Link",
			"options": "UOM",
			"width": 115,
		},
		{
			"label": _("Delivery Note"),
			"fieldname": "delivery_note",
			"fieldtype": "Link",
			"options": "Delivery Note",
			"width": 130,
		},
		{
			"label": _("Project"),
			"fieldname": "project",
			"fieldtype": "Link",
			"options": "Project",
			"width": 120,
		},
		{
			"label": _("Company"),
			"fieldname": "company",
			"fieldtype": "Link",
			"options": "Company",
			"width": 120,
		},
	]


def get_data(filters):
	conditions = get_conditions(filters)

	data = frappe.db.sql(
		f"""
			SELECT
				dni.item_code,
				dni.item_name,
				dni.item_group,
				dni.brand,
				dni.uom AS uom_qty,
				dn.customer,
				dn.customer_name,
				dn.customer_group,
				dn.territory,
				dn.project,
				(SELECT GROUP_CONCAT(sales_person SEPARATOR ', ') FROM `tabSales Team` WHERE parent = dn.name) as sales_person,
				dn.posting_date,
				dni.qty,
				{STOCK_ORDERED_QTY_SQL} AS stock_ordered_qty,
				{STOCK_DELIVERED_QTY_SQL} AS stock_delivered_qty,
				COALESCE(bin.actual_qty, 0) AS stock_available_qty,
				item.stock_uom AS production_uom,
				GREATEST(
					{STOCK_ORDERED_QTY_SQL}
						- {STOCK_DELIVERED_QTY_SQL}
						- COALESCE(bin.actual_qty, 0),
					0
				) AS stock_shortfall_qty,
				dn.name as delivery_note,
				dn.company
			FROM
				`tabDelivery Note Item` dni
			INNER JOIN
				`tabDelivery Note` dn ON dni.parent = dn.name
			INNER JOIN
				`tabItem` item ON item.name = dni.item_code
			LEFT JOIN
				`tabBin` bin ON bin.item_code = dni.item_code AND bin.warehouse = %(warehouse)s
			LEFT JOIN
				`tabSales Order Item` soi
					ON soi.name = dni.so_detail AND soi.parent = dni.against_sales_order
			WHERE
				dn.docstatus = 1
				{conditions}
			ORDER BY
				dn.posting_date desc
		""",
		filters,
		as_dict=1,
	)

	return data


def get_conditions(filters):
	conditions = ""
	if filters.get("from_date"):
		conditions += " AND dn.posting_date >= %(from_date)s"
	if filters.get("to_date"):
		conditions += " AND dn.posting_date <= %(to_date)s"
	if filters.get("company"):
		conditions += " AND dn.company = %(company)s"
	if filters.get("customer"):
		conditions += " AND dn.customer = %(customer)s"
	if filters.get("customer_group"):
		conditions += " AND " + get_descendant_condition(
			"Customer Group", "dn.customer_group", "customer_group"
		)
	if filters.get("territory"):
		conditions += " AND dn.territory = %(territory)s"
	if filters.get("project"):
		conditions += " AND dn.project = %(project)s"
	if filters.get("item_code"):
		conditions += " AND dni.item_code = %(item_code)s"
	if filters.get("item_group"):
		conditions += " AND " + get_descendant_condition("Item Group", "dni.item_group", "item_group")
	if filters.get("brand"):
		conditions += " AND dni.brand = %(brand)s"
	if filters.get("sales_person"):
		conditions += " AND EXISTS (SELECT name FROM `tabSales Team` WHERE parent = dn.name AND sales_person = %(sales_person)s)"
	if cint(filters.get("pending_only")):
		conditions += f" AND {STOCK_ORDERED_QTY_SQL} > {STOCK_DELIVERED_QTY_SQL}"
	return conditions


def set_default_warehouse(filters):
	if filters.get("warehouse") or not filters.get("company"):
		return

	filters["warehouse"] = frappe.db.get_value(
		"Warehouse",
		{
			"company": filters.company,
			"warehouse_name": "Finished Goods",
			"is_group": 0,
			"disabled": 0,
		},
		"name",
	)


def validate_warehouse(filters):
	if not filters.get("warehouse"):
		frappe.throw(
			_("Please select a Warehouse. No enabled Finished Goods warehouse was found for {0}.").format(
				filters.company
			)
		)

	warehouse = frappe.db.get_value(
		"Warehouse",
		filters.warehouse,
		["company", "is_group", "disabled"],
		as_dict=True,
	)
	if not warehouse or warehouse.disabled or warehouse.is_group:
		frappe.throw(_("Please select an enabled, non-group Warehouse."))
	if warehouse.company != filters.company:
		frappe.throw(_("Warehouse {0} does not belong to company {1}.").format(filters.warehouse, filters.company))
